PurchasesCoreSwift/Public/EntitlementInfo.swift

```diff
@@ -0,0 +1,346 @@
+     billing issue or an issue has been resolved.
+
+     Note: Entitlement may still be active even if there is a billing issue.
+     Check the `isActive` property.
+     */
+    public private(set) var billingIssueDetectedAt: Date?
+
+    /**
+     Use this property to determine whether a purchase was made by the current user
+     or shared to them by a family member. This can be useful for onboarding users who have had
+     an entitlement shared with them, but might not be entirely aware of the benefits they now have.
+     */
+    public let ownershipType: PurchaseOwnershipType
+
+    // TODO(post-migration): Make this internal
+    // TODO(cleanup): Codable
+    public init(entitlementId: String,
+                entitlementData: [String: Any],
+                productData: [String: Any],
+                dateFormatter: DateFormatter,
+                requestDate: Date?) {
+        // Entitlement data
+        let entitlementExpiresDateString = entitlementData["expires_date"] as? String
+        let entitlementPurchaseDateString = entitlementData["purchase_date"] as? String
+        let productIdString = entitlementData["product_identifier"] as? String
+
+        // Product data
+        let periodTypeString = productData["period_type"] as? String
+        let originalPurchaseDateString = productData["original_purchase_date"] as? String
+        let productExpiresDateString = productData["expires_date"] as? String
+        let storeString = productData["store"] as? String
+        let maybeSandbox = productData["is_sandbox"] as? NSObject // This could be a String or NSNumber
+        let unsubscribeDetectedAtString = productData["unsubscribe_detected_at"] as? String
+        let billingIssuesDetectedAtString = productData["billing_issues_detected_at"] as? String
+        let ownershipType = productData["ownership_type"] as? String
+
+        let store = Self.parseStore(store: storeString)
+        let expirationDate = Self.parseDate(dateString: productExpiresDateString, withDateFormatter: dateFormatter)
+        let unsubscribeDetectedAt = Self.parseDate(dateString: unsubscribeDetectedAtString,
+                                                   withDateFormatter: dateFormatter)
+        let billingIssueDetectedAt = Self.parseDate(dateString: billingIssuesDetectedAtString,
+                                                    withDateFormatter: dateFormatter)
+        let entitlementExpiresDate = Self.parseDate(dateString: entitlementExpiresDateString,
+                                                    withDateFormatter: dateFormatter)
+
+        self.store = store
+        self.expirationDate = expirationDate
+        self.unsubscribeDetectedAt = unsubscribeDetectedAt
+        self.billingIssueDetectedAt = billingIssueDetectedAt
+        self.identifier = entitlementId
+        self.productIdentifier = productIdString!
+
+        let isSandbox: Bool
+        if maybeSandbox?.responds(to: #selector(getter: NSNumber.boolValue)) ?? false,
```

Comment:
Here's a fun one, NSNumber and NSString have the same `-boolValue` method so in ObjC the message passing worked regardless of type because types are just a suggestion anyways, right? In swift, I had to decide what type it was. First I was like "yeah, this is probably a string" but that didn't work. I'm not going to assume it's an NSNumber or an NSString, so I'll just check for `NSNumber` first (because that's the most likely and I like to pre-optimize).

Reply:
added a comment before reading this. I think we should be able to figure out its concrete type since it's coming from our own rest api

Reply:
Just asked in our company chat.